### ISSUE TITLE
No sound - Creative Labs Sound Blaster Recon 3D / Z (ca0132)

### ISSUE THREAD

**Issue by ThEMarD** [ValveSoftware/SteamOS#221]
Heyyo,

My Creative Labs Sound Blaster Recon 3D (ca0132, same DSP as the Sound Blaster Z) can't output sound or record from my microphone.

I'm running SteamOS with Alchemist_beta fully updated.

I tried the fixes that worked when using Arch Linux or Ubuntu of editing the /etc/rc.local and adding in...

rmmod snd_hda_intel
modprobe snd_hda_intel position_fix=1

or...

sudo kill `ps uax |grep pulseaudio | grep -v grep | awk '{print $2}'` ; sudo rmmod snd-hda-intel; sleep 3; sudo modprobe snd-hda-intel

and neither worked... all I get is static. I'm unsure if this is kernel related (since ca0132 workaround has worked in kernel 3.10+ when I've tested it) or something that SteamOS changes compared to other Linux distros.

Currently, Creative Labs have not put out a proper Linux driver for any ca0132... the current support for ca0132 sound cards comes from porting their official Google Chromebook Pixle drivers to ALSA & the Linux Kernel.

here's my alsa-info:
http://www.alsa-project.org/db/?f=8aefb0663de38f0e3e2f1df4facd257aaebbb3bc

I hope this helps. Please let me know if there's anything further I can try to help resolve this issue and hopefully get SteamOS working with the ca0132 based sound cards for all users. :)

**Comment by jadbox** [comment]
Same here, I really wish I had my SBZ working under SteamOS...

**Comment by martinmolin** [comment]
ZxR here, also broken in 64bit Linux (ca0132).. currently the only thing stopping me from dropping Windows. I haven't seen any positive response from Creative themselves when "regular" people pokes them for Linux drivers but I'd guess if they see that it's worth implementing it, they'll do it.. So lets hope SteamOS is a success, even more so than what it already is.

**Comment by itsexe** [comment]
I have the same problem. This is the only reason I use windows.

**Comment by SwaginWagon101** [comment]
Yep. Same Problem. This is a real bottleneck for me....

**Comment by Grief** [comment]
+1

**Comment by Grief** [comment]
I guess this won't be fixed...

**Comment by Tele42** [comment]
Creative have expressed their stance by their actions, once upon a time they supported linux, but have not in many years. Get this sound card working on linux in general, and it will trickle down to SteamOS.

**Comment by ProfessorKaos64** [comment]
Is there source code released for the Chrombook variant of this firmware? Is it x86 or ARM? Can the module be rebuild under a test chroot or test machine? If I had the hardware, I'd be more interested in trying to help out. What steps do you use to make this work in Debian 7/8? Is there any particular reason you are running Alchemist over Brewmaster? You will have much newer packages, and systemd if you upgrade. It is worth updating if you are adamant about getting this working.

**Comment by ThEMarD** [comment]
Heyyo, @ProfessorKaos64 , I don't think Brewmaster was out in Aug 17, 2014 when I made this topic. I can try installing it sure, but the latest 64bit kernel v4.2.0 has been giving me the same grief as whatever 64bit kernel v3.x  that alchemist_beta used. Tested with Linux Mint, Manjaro and Ubuntu Mate.

The issue seems to be the 64bit kernel firmware loader or the firmware files themselves aren't 64bit compatible... ctefx.bin ctspeq.bin. It's included in many Linux distributions and can even be found I think with a quick Google search but no matter where I put the files they never load in 64bit and require that rmmod and modprobe terminal command to try and get working. I'm not sure how that circumvents the 64bit kernel firmware loading issue but it does... I kinda wish I picked up coding but I never found the time I guess... if you think you can do something with the ctefx.bin and ctspeq.bin files to make them load under a 64bit kernel that would be GREATLY appreciated.

I made this comment thread to hopefully get Valve to tell Creative to do something about their lack of Linux/SteamOS support and hopefully fix it but haven't had much luck yet sadly.

You can see the original Linux kernel bug report here from over two years ago:
https://bugzilla.kernel.org/show_bug.cgi?id=55541

**Comment by jadbox** [comment]
Sound Blaster Z support is one of my last remaining wishes for Debian/Linux. Anyone want to start a bug bounty pool on this?

**Comment by ThEMarD** [comment]
Heyyo, someone's been working on it! Takashi Iwai!!! There is hope! https://bugzilla.kernel.org/show_bug.cgi?id=55541

**Comment by martinmolin** [comment]
Oh wow. You just made my day ThEMarD. Thank you!!

**Comment by jadbox** [comment]
Sweet mother of God... I nearly gave up watching this issue. As a linux gamer, my excitement is unbounded!

**Comment by ThEMarD** [comment]
Heyyo, The fix is now part of Kernel 4.4-RC1 for testing! Hopefully it gets approved and passed down to a stable kernel. Takashi Iwai is a very big hearted and humble gentlemen. He declined my PayPal donation too saying he's just happy to help so be sure to test out that kernel and report back to that bugzilla link in my previous post if it doesn't work for you.

**Comment by martinmolin** [comment]
This is awesome news. I'll have to test this, much love to Takashi Iwai!

**Comment by voron00** [comment]
We are trying to get it fixed. See here https://bugzilla.kernel.org/show_bug.cgi?id=109191
Basically, the DSP firmware is incomptabile with SB Z-Series. Jaroslav Kysela is still trying to do something with it but i feel somewhat sceptic about that. Anyway, there is a way to make it work thanks to Takashi Iwai and his patch to use generic hda driver for SBZ. I've updated this patch to use proper pins for SBZ (still needs testing on Zx | ZxR) for Line in (rear mic) and Line out (3rd jack) (Yeah, only line out is working at the moment and mic). My patch is here https://github.com/voron00/linux/commit/fcbf5ba088c9bc8104f32738e356600ecfb85689
You may try it yourself, just compile the kernel from my repo or cherrypick the patch and compile with your own kernel.

**Comment by Millnert** [comment]
I had sound for a while but since debian 4.6.1 on debian stretch and recent package upgrade (2016-06-20) , it stopped working again. Now trying the patch at https://bugzilla.kernel.org/show_bug.cgi?id=109191#c14 , which applied cleanly to debian 4.6.1 source. I had one change in my PCI ID which is 1102:0011 rather than the 1102:0010 the patch was written for first. The motherboard is a Gigabyte Z170X-Gaming 7.

**Comment by NukeMania** [comment]
+1 still it's not working
https://bugzilla.kernel.org/show_bug.cgi?id=120491

**Comment by sweca** [comment]
2018 and still not working. Creative don't care.

**Comment by Conmanx360** [comment]
Hi!
I've been working on my own patch, and am currently in the process of submitting it to the Linux Kernel. In the meantime, if anyone wants to compile their own kernel and test the early version, you're welcome to. 
Here's the thread I posted on the Creative forums:
http://forums.creative.com/showthread.php?t=742256

If you don't want to read that stuff, you can just download it straight from my Google drive:
https://drive.google.com/drive/folders/1r-X8g6It5AO7SwgJc1JQfJ8KCOyqrr_d

Put the patch_ca0132.c in your kernel source's sound/pci/hda folder, and put the firmware ctefx-sbz.bin in your /lib/firmware folder. Let me know if it works for you. I'm still trying to find people with the variants of 0x0024 and 0x0025 to test it.

Thanks.

**Comment by medzernik** [comment]
> I've been working on my own patch, and am currently in the process of submitting it to the Linux Kernel.

What does that patch do? Does it enable new features?

**Comment by mirh** [comment]
It enables the card to work properly, so.. yes I guess?

It has been already merged in linux-git (and for who's bound not to be able to compile a kernel itself it'll be official in 4.18)

**Comment by Conmanx360** [comment]
Yes, it loads the firmware, so you don't have to boot into Windows first. It also allows you to mess with the effect levels. Here's what the new alsamixer looks like:
https://drive.google.com/drive/folders/1FGaLli4rKrVmBxpx0ulGF3k4ISJ3JCYj

**Comment by medzernik** [comment]
That looks phenomenal! Absolutely phenomenal!! so I can finally put my recon3D card back in!!

**Comment by Conmanx360** [comment]
I believe I could get it to work on the new AE-5, which has the same Core3D onboard. The only thing is, I need the data, which someone could get by running my program QemuHDADump. It dumps all of the commands from Windows, which is how I've gotten the Sound Blaster Z and Recon3Di to work. So, basically, I'm pretty positive I could get any Core3D based card to work, I just need someone who owns the card to get me the data. That's the only thing keeping me from supporting them.

**Comment by medzernik** [comment]
I will purchase the card and do as you say. Feel free to contact me further here or on my email: medzernik1@gmail.com

**Comment by Crypher19** [comment]
This is really great, thank you for your firmware. It works perfectly with my card SB-Z (0x0023 variant), though I have two questions:
1) After kernel 4.18 installation will still be necessary to put the firmware ctefx-sbz.bin in /lib/firmware folder?
2) Since Creative shut their forum down, @Conmanx360 could you please repost the topic anywhere else? (maybe here on github, your wiki?).

Thanks.

**Comment by Conmanx360** [comment]
I recently got an email from Creative saying that I had permission to submit the firmware in the same way the Chromebook's current ctefx.bin firmware is included, I haven't gotten around to getting that done, as I suspect it'll require some correspondence with Creative again to iron out the details. I do intend on putting it in a Wiki article, and including a guide on what I need in a dump for people who want support for their cards. 

The Wiki part shouldn't take me too long, as I've already written a lot of the stuff previously. The firmware stuff, I have no clue yet. Need to do more research. But, the firmware for the Chromebook works, I just don't know if it makes any changes.

**Comment by mirh** [comment]
Couldn't that be sent to linux-firmware@kernel.org ?

**Comment by Conmanx360** [comment]
It's been awhile, but I recall reading that you need some sort of licensing to be sorted out before sending anything to that address. I almost think Creative would have to send it themselves, I don't work for them and have no involvement with them so I don't believe I could send something under a license on their behalf with just an e-mail as proof they gave me permission.  

They're not exactly super responsive, so getting all of this out of them may take some time. Sometimes they just won't respond for weeks. It seems like the best way to get a response is to say bad things about them on Reddit, which happened coincidentally the same day they sent me an email saying I had permission to do it. I used to be able to bug the moderator on their forums to get them to respond, but that is no longer an option.

**Comment by Jacek13** [comment]
@Conmanx360 Thanks for your work, but on kernel 4.18.0-041800rc3-generic I can't get 5.1 setup to work, all channels other than stereo still play by both stereo speakers at the same time. In alsamixer it properly says Recon 3Di. I did not install your drivers, just using those integrated in the kernel.
If you could do a bit more magic, I'd be really grateful!

**Comment by Conmanx360** [comment]
The ones in 4.18 are the ones that I wrote. Did you try muting "HP/Speaker Auto Detect" and setting "Output Select" to surround? Also, make sure you set pavucontrol to 5.1. People have had mixed results with surround on the Recon3Di, and unfortunately I don't own one to test on myself. Try what I suggested and let me know.

**Comment by dsteidley** [comment]
I would like to get involved with this as well.  I have not only just found this thread, but I have also recently added a SB Recon3d (*no i*).  I will send you an email to touch base, and then I will send all of my specs.  While I have used Ubuntu for a while, I am by no means an expert.  I guess that is why I want to stick this one out... just to say I learned something.  I am also willing to help with the documentation.

**Comment by QDesjardin** [comment]
I have the AE-5 card and drivers -- if nobody has done it yet, how can I use QemuHDADump to dump the firmware for Linux?

**Comment by Conmanx360** [comment]
https://github.com/Conmanx360/QemuHDADump/wiki/Setup-and-usage-of-the-program Is the guide I wrote to use it. Let me know if you have any issues.

**Comment by QDesjardin** [comment]
So I'm running Arch Linux, have the AE-5 passed through VFIO, and I already have an existing Windows 10 VM that was setup through virt-manager. The thing is when I compiled that QEMU from source, and tried to launch my VM from there, my whole system froze. 

I'm wondering, assuming that Arch's QEMU already has tracing support compiled in, how can I launch my existing VM so that it does the tracing?

**Comment by Conmanx360** [comment]
Arch's QEMU probably doesn't have tracing built in, due to it slowing things down. What are your command line options for qemu on your existing VM? You may arguments set that aren't set for when you try to run QemuHDADump, which could cause crashes due to incompatibility.

**Comment by QDesjardin** [comment]
What I'm going to do actually is create a new Windows 10 VM from scratch with the compiled QEMU, and attach the AE-5 card onto there. If that works, I'll send you the dump.

**Comment by QDesjardin** [comment]
My terminal is outputting the stuff now, and it's making frameXX files. 

![screenshot_2018-07-16-12 49 56](https://user-images.githubusercontent.com/22847899/42777587-cb91da30-88f7-11e8-94a7-a5f94b820558.png)

What do I do exactly so everything's captured, because [this guide](https://github.com/Conmanx360/QemuHDADump/wiki/For-Creative-cards:-How-to-send-me-a-dump) is a bit iffy to comprehend.

**Comment by Conmanx360** [comment]
Yeah, guess it isn't super clear. Basically, toggle through all the stuff listed. Turning the output effects on/off, switching between surround, headphone, stereo, and then doing the same with all the possible mic inputs. Write the verb number down before you do the action though. So, what I do, is when Window's has loaded and there are no more 'verbs' (verbs are just the address of the verb in the file I create) scrolling, write down what the last one was. 

That verb can be labeled something like 'Startup complete.' Then, toggle output effects. After that's done, mark down what the last verb was, and label it 'output effects end.' That's the gist of it. Like I said earlier, you'll want to do this for all the different output options (surround, stereo, headphone) and input options (front mic, rear mic, rear line-in) and also the input effects. Then, you should send me all of those verb locations, as well as the console output in a text file to my email conmanx360@gmail.com . Hopefully that clears things up a little.

**Comment by QDesjardin** [comment]
QemuHDAdump crashes upon loading into Windows 10 VM, it's not taking down the verbs. It seemed to be working fine before when I first tried it.

![screenshot_2018-07-16-14 57 28](https://user-images.githubusercontent.com/22847899/42783267-b340cc8c-8908-11e8-826e-0be3e8b1f53b.png)

**Comment by Conmanx360** [comment]
That looks like it's accessing memory areas it shouldn't be in. Try restarting. For what it's worth, other people have had more issues with the program than I have... It looks like it began to take down verbs, then crashed after the first frame dump.

**Comment by QDesjardin** [comment]
I tried restarting - it crashes just like this again.

**Comment by Conmanx360** [comment]
Hmm, did you clear the previous frame files? Are you using the right pty? Are you running as sudo?

**Comment by QDesjardin** [comment]
I cleared the previous frame files, running in the right tty (unless "pty" is something else) and under sudo, yes.

**Comment by Conmanx360** [comment]
Could you possibly join our IRC channel, #ca0132 on freenode? It'd be a lot easier to troubleshoot that way.

**Comment by Conmanx360** [comment]
Okay, we managed to get an AE-5 dump, and I have a Recon3D being sent in the mail for me to test on. Getting to work immediately. I used to set time tables, but after not being able to ever get them right, I don't think I'll do that anymore. I'll find a way to get it out when I've got them working to find testers.

**Comment by medzernik** [comment]
I am finally not sick anymore, though it will be on Monday that I will
finally be available. I apologize for not getting the dump in time,
however, I can serve as a tester for the card when you have the prototype
driver done.

**Comment by Conmanx360** [comment]
Just got the Recon3D working. I'm going to do more testing to make sure it's stable, and then I'll get it split into patches to try to get accepted into the kernel. Not sure which card I'll work on next. I've got a ZxR dump as well as an AE-5 one. There seems to be more people willing to test the AE-5, so I guess I should move onto it. Probably going to need lots of tests, as it seems to be kind of unique.

Anyways, I'll let you guys know when I make more progress.

**Comment by Conmanx360** [comment]
Okay, just sent the Recon3D back to it's owner, I think I'm going to begin work on the AE-5. Hopefully it doesn't take too long, however, I will need people to test the AE-5 when I've got most of it written. Currently, there are some writes to PCI region2 that I believe to be related to the LED's on the card, but I'm not sure. I guess we'll find out once it's tested.

Without one on hand, testing is going to take longer. I'll update here when I've made progress and am ready for testers. Feel free to join freenode channel #ca0132 and ask questions if you have any.

**Comment by QDesjardin** [comment]
As far as I remember, the region2 writes have nothing to do with the LED lights when changing the acoustic engine settings.

**Comment by Conmanx360** [comment]
Hm. Might be worth it to just change the LED colors and check. I think certain modes also have effects on the LED's. If it turns out all the region2 writes are all audio related, there's potentially going to be a lot of testing needed. Tomorrow I'll get into looking at it a bit deeper. The only thing that makes me believe it's LED related is that there are still writes to addresses that are used for GPIO on the Recon3D and Sound Blaster Z, so I don't believe the different addresses are related to that.

Plus, the LED stuff has to be controlled somehow, and it definitely isn't through HDA verbs. But, we'll find out. Tomorrow I'll start writing stuff into the driver.

**Comment by Conmanx360** [comment]
Recently, I had someone send me their Recon3D, I got the data I needed, added support, and then sent it back. If anyone inside the United States would like to do that with the ZxR or AE-5, I will pay for the return shipping. It would speed the process along by quite a bit. If you're interested in doing this, email me at conmanx360@gmail.com .

The only reason I mention this, is because it looks like both cards have a few oddities that I think will require a fair bit of messing with to work out, and it goes a lot faster when I have the actual hardware. I would buy the cards myself, but they're a bit expensive, and I don't think I'd ever use them once I was done with them.

**Comment by Conmanx360** [comment]
Seems weird to keep updating here, but I really have nowhere else to do it.

Anyways, got the Recon3D stuff merged into ALSA. Now working on the AE-5. Alastair Bridgewater and I (he did most of the work) have been disassembling the program memory of the ca0132's integrated 8051, which has been helping us to understand some of the stuff going on internally. Only reason I mention this is, it may slow me down a bit, as I'm trying to understand some of the stuff I'm putting into the driver rather than just straight copying the verbs.

So, AE-5 stuff is now being worked on. Don't have a time table. Will update when I have made progress.

**Comment by ThomasAH** [comment]
* Connor McAdams <notifications@github.com> [20180809 01:59]:
> Seems weird to keep updating here, but I really have nowhere else to do it.

This is the best thread I have found about this problems, and few
days later (early June) you wrote that you're working on a new
patch, so I think this still is the best place to keep us updated.

Thank you for all your work!

I recently bought a Gigabyte GA-Z170X-Gaming 7 with an on-board
Recon3Di, so I'm happy that I soon will have sound again.

> Anyways, got the Recon3D stuff merged into ALSA.

I'm currently using Debian stretch on amd64 with a 4.17 kernel.
Does someone here have a recommendation about the best way to test
the new drivers?

Debian experimental has a linux-image-4.18.0-rc4-amd64 package, does
this already contain the latest changes or is a newer rc needed?

(huh, is it already 10 years since the last time I had to compile a
kernel myself for a regular PC?)

And besides the kernel and ctefx-r3di.bin firmware file, do I need
anything else?

Regards,
Thomas

-- 
Thomas Arendsen Hein <thomas@intevation.de> - OpenPGP key: 0x5BB3F5195816791A
https://blogs.intevation.de/thomas/         - https://intevation.de/~thomas/
Intevation GmbH, Neuer Graben 17, 49074 Osnabrueck - AG Osnabrueck, HR B 18998
Geschaeftsfuehrer: Frank Koormann, Bernhard Reiter, Dr. Jan-Oliver Wagner

**Comment by Conmanx360** [comment]
I'm pretty sure your card is included in the driver, if you could post your lspci -vv -nn I could tell you definitively. There are quite a few different Recon3Di's.

**Comment by Jacek13** [comment]
@ThomasAH I have exactly the same motherboard (Gigabyte Z170X Gaming 7), and all you need to use the current driver is the 4.18 kernel, no other files necessary. And does the driver works with it... Kind of. It works perfectly, but only with stereo. If you set it to surround setup, like 5.1, the sound still comes just from stereo speakers. What I discovered is no matter the settings the stereo output jack is configured as line out, and all other are not assigned. I have already contacted @Conmanx360 about it a few days ago.
BTW, @Conmanx360 could you share about how did you learn how to reverse engineer and write drivers for Linux? I'm very interested in that. Did you read any good books about it?

**Comment by Conmanx360** [comment]
Yeah, I haven't looked into that issue yet, but I'm suspecting it is probably a GPIO issue.

Anyways, there wasn't really a straightforward path... It was kind of just following one lead after another. When I started, I hadn't written anything other than some simple Arduino stuff. I started out reading the Intel HD Audio spec sheet, which I think most people who came to this problem did, only to realize that the ca0132 doesn't really pay much attention to HDA stuff once the DSP is loaded. So, I figured I would have to see what commands were being sent to it in Windows. Found in the HD Audio spec sheet that commands, or verbs as they're called here, are sent to the card through a DMA buffer. Then, I had to find out how DMA worked, how to get access to that buffer, and it just kept going lower level until I had actually found the buffer, then made a program to get the virtual machine to dump that data area for me. I'm no expert at reverse engineering, but I'd say the biggest part of it is just learning how the thing you're trying to reverse engineer works, and then finding someway to watch it's behavior. 

As far as books, I don't think I read any on reverse engineering in particular, more so just general computer books. ldd3, Linux Device Drivers 3rd Edition is the one I read to understand how modules work, and I read some college book I found in pdf form on low level computer stuff. That really helped. Oh, and also, a book called Understanding The Linux Kernel 3rd Edition. Only one I read from front to back was the college book I found, but the others I just read what I wanted to understand at the time, which was mainly DMA related stuff.

I wrote a post on reddit about it here: https://www.reddit.com/r/linux/comments/8rlehm/beginners_to_contributing_to_the_linux_kernel/e0ttvqd/?context=3

Anyways, hope that helps. Also, the links to the firmware:

ctefx-r3di.bin: https://drive.google.com/open?id=1hvEweGHzlBpAeaJvvQRIi-utq01zbl-v

ctefx-sbz.bin: https://drive.google.com/open?id=12qZTWTkWA_BE1e4iXCQf3-l-gQI59E5F

**Comment by Conmanx360** [comment]
Went ahead and bought an AE-5 on Amazon. I don't really have a use for it, so hopefully I'll be able to resell it once I'm done.

**Comment by medzernik** [comment]
I am sure if you'd set up donations, you'd get some too. There are a lot of people who really really want to get the card working in Linux.

**Comment by Conmanx360** [comment]
Some further digging into it, if the card has already booted into Windows once, it really enjoys spamming startup commands. It does it so fast it overloads QemuHDADump. So, only one initial shot to get a dump, not impossible, but just... odd behavior.

Also, the LED effects, like for instance making the lights onboard pulse. It's not a thing where you send the card a command saying "pulse the LED's" it's actually the PC constantly telling the card what to do. Seems... wasteful. I dunno. Maybe Creative knows better than I do. Anyways, interesting stuff.

**Comment by Conmanx360** [comment]
Found a good deal on the ZxR on eBay, so I went ahead and bought one. So, support should be easier to get once I start working on it. I've got the command structure for the AE-5's region2 writes down, now I just need to break it down into different behaviors, i.e this command does headphones etc. Should (hopefully) have some sound out of it by the end of the week.

**Comment by Conmanx360** [comment]
Took a bit of a detour to do some 8051 disassembly to try and figure out what each ParamID does. Managed to find the table of handler addresses for all of the ParamID's, so this should give me the ability to identify what  the ones on the AE-5 that aren't currently defined actually do. Maybe this will help clear some things up.

Don't want anyone to get the impression I've given up on it, working at it every day.

**Comment by Conmanx360** [comment]
@AvdWerfhorst - I understand your pain. Windows drivers suck, and with the Linux drivers being built into the kernel, that should eliminate that issue in Linux. I'll update here once I get into ZxR stuff. There seems to be quite a few people waiting on that.

Now, onto AE-5 stuff: Currently trying to work out the weird region2 command sequences. All I want is to hear the freaking headphone relay click. I think I've realized what's wrong, I'll have to read certain registers after writing them to clear them. At least, that's the current theory. I'll find out tomorrow, too tired to keep going at it tonight. However, this is making me believe the ZxR would've been easier to work with, even with it's weird dual codec thing. Doesn't really matter now, just thought it was funny that the card I thought would be easier to write a driver for is seemingly the more complicated one. Guess I'll find out once I get to work on the ZxR. Anyways, slowly getting there. Still hoping to get sound out of the AE-5's rear ports by at least the end of the month. That's all I have to report for now. Back to work.

**Comment by Conmanx360** [comment]
Okay, got the headphone relay clicking from within Linux. That's a pretty good step. Now, onto the bigger task of actually getting the rest of it working.

The LED stuff seems to be kind of complicated. I may hold off on getting support for it until after I've got the ZxR working, just for the sake of time spent on things that are necessary vs aesthetic.

**Comment by medzernik** [comment]
the LEDs are a mess to work with even on Windows, can't imagine working with it from a driver standpoint. Personally, the most important I think is that you can switch between headphones/5.1 audio, switch the audio configurations, use the different options and effects and the EQ working. It's great it's progressing!

**Comment by Conmanx360** [comment]
@medzernik - As far as the LED's are concerned, yeah, they're a mess. I've thought of a few ways to give access to them from some sort of external program, but haven't tested any of it yet or really looked too deep into it. But, I'd like to have support for at the least switching colors, even if there's no pulsing or anything.

**Comment by medzernik** [comment]
@Conmanx360 I mean, the entire card is a mess :D the bugs on Windows and the control panel and how it's shit you have no idea... Sometimes there is an exception and the card just fries off the sound, putting 100% volume into your speakers/headphones with a crackling 8 bit sound. Sometimes the control panel doesn't remember settings. Sometimes it does. Sometimes the EQ resets. It's a mess.

**Comment by Conmanx360** [comment]
@medzernik  - Hm... well, hopefully the Linux driver turns out to work better. I know I don't have many issues with the Sound Blaster Z, but I don't recall having TOO many issues in Windows, although I have heard of other people having issues. There should just be less 'mess' in Linux, because there's no bloated GUI to deal with. But, we'll find out once I get it working. Too many factors to tell just yet.

**Comment by medzernik** [comment]
@Conmanx360 haha I am certain it will work better :D unless pulseaudio decides to f-out. I hope I can finally get a reliable equalizer in Linux. It's almost in reach! :D And with the recent Steam news of new games, this will be like second christmas :D

**Comment by Conmanx360** [comment]
aaaaaaaand I've now got sound out of line-out working. Okay. Good stuff.

Currently, no sound out of the headphone port though, I'm kind of suspecting I'm not turning on the amplifier or something, because in Windows, there's a hiss when they're connected. Not hearing that in Linux, so my early guess is that it must not be on. 

Well. We're now one step closer to full AE-5 support. I guess I can post the current patch_ca0132.c, but keep in mind, it's a friggen mess. The mic seems to have the Sound Blaster Z issue on my end, where it can be intermittent. It's fixable by toggling inFX on and off quite a few times. That's something that 8051 disassembly should help me resolve, but it doesn't seem to effect everyone. Anyways.

Link to patch_ca0132.c for the AE-5 is here: [Link to Google Drive folder](https://drive.google.com/open?id=1JTo9lGBYLJ5JBP9yBBhqbBfQrjchf0fX)

I will upload newer versions as they come there, in the same way I did with the Sound Blaster Z. Eventually, I should get it cleaned up enough for acceptance into the kernel, and then you won't have to download it anymore. Also, the AE-5 uses the Sound Blaster Z's firmware, link is here: [ctefx-sbz.bin](https://drive.google.com/open?id=12qZTWTkWA_BE1e4iXCQf3-l-gQI59E5F)

If you wish to donate, link is here: [Donate](paypal.me/conmanx360)

Okay. That's enough for now. I will update when I get the headphones and stuff working, and when I get new versions uploaded.

**Comment by Conmanx360** [comment]
Fixed front headphone. Will update tomorrow with that and shutdown stuff (probably).

**Comment by medzernik** [comment]
Also if possible, please add everything to the alsamixer controls. It's extremely important to control all the functions :D Otherwise, exciting and great news coming! I will be sure to donate next month when I get some spare money.

**Comment by Lolzen** [comment]
Really enjoying the Sound the soundcard is putting out now, thanks to you!
However, i do have problems getting the mic to pickup anything. Even inFX toggle does nothing except some occasional "PLOP". Double checked, i have it plugged in in the Mic In jack.
Headset does work, as it is rather new, and also works on onboard which has it own problems. That's especially frustrating for me personally as i can't use voice programs to communicate which is the only thing holding me back from wiping my Win7 install now.

**Comment by Lolzen** [comment]
Thank you, i didn't mean to sound pushing rather than giving feedback. I'm just hyped and thrilled at the thought that almost nothing is blocking me from going full ext4 in the near future.
To be honest i'm glad the problem isn't due to my end, i thought i might have messed something up as you stated it works with minor hiccups, so thanks for confirming and giving me relief!

Also, if you need more extensive testing for anything let me know if i can somehow help.

**Comment by Conmanx360** [comment]
@Lolzen - It didn't come across that way :) It's a problem that's alluded me for awhile with the Sound Blaster Z, and seems to also affect the AE-5. Everything used to work just fine until I cleaned up the code. Then, it stopped working. I've looked at the diffs, had other people look at the diffs, no command changes. Like I said, my theory is a timing issue, and it may require some 8051 disassembly to understand what in the heck is going wrong with it. It's solvable, I'm just not sure how much work it will be, which is what makes it hard to give an answer as far what testing will have to be done/when it will be done.

For now, I'm going to focus on getting the AE-5 code cleaned up into a state where I can send it off as patches, and then I will begin work on the ZxR, which from some more recent dumps that I've gotten, seems to be less difficult than I previously believed. It seems like it will be able to re-use a lot of the Sound Blaster Z's code, which is always useful. The only thing that will be an issue with it is the daughter board. I haven't done too much digging, but I'm starting to think it may be controlled through some sort of weird audio stream interface. Not 100% sure on that, just my current theory. Once that's sorted out, then I will probably come back around to the microphone issue.

It's crazy, but we're pretty close to having all of the Creative ca0132 cards functioning without Creative's help.

**Comment by Ashalyss** [comment]
@Conmanx360 I am an owner of the AE-5.

I am running Manjaro Gnome with an untouched kernel 4.18 and currently I get microphone input from the rear port and headphone output from the front ports only. This is only after using alsa mixer to unmute the detection and the HP port.

In Windows I get full functionality.

I am very interested in your work as I would really like to get this card working properly from the rear ports.

I am also a bit of a linux newbie and I don't know how to apply patches. Do you think your patches will be merged with the mainline kernel at some point? If not would you be able to tell me how to make your patches work?

**Comment by Conmanx360** [comment]
@Ashalyss - To test these, you'll need to compile your own kernel. It should eventually get merged, but that may be awhile. There may be a way to compile it for your current kernel on Manjaro using DKMS, but I haven't done too much work wtih that. There's a thread here: https://forum.manjaro.org/t/how-to-patch-manjaro-linux-kernel-for-sound-blaster-z/48559/14 that has someone doing it for the Sound Blaster Z, all you'd have to do is replace that patch_ca0132.c with the one I linked above.

**Comment by Conmanx360** [comment]
Okay, got the ZxR working. Only tested headphone out and microphone port, both seem to work fine. Uploaded the patch_ca0132.c to the Google drive link posted up above. More cleanup to be done with it, and some other misc stuff to do, but getting pretty close to done. ZxR users, please let me know how it works.

**Comment by Lolzen** [comment]
recompiled the kernel with the new patch and can confirm mic now works beautifully along with sound on the AE-5. Thank you!
I'm impressed by the progress in such a short amount of time.

**Comment by medzernik** [comment]
So the sound now works on AE-5? Are controls for the sound also available in alsamixer?

**Comment by Conmanx360** [comment]
Oh! I didn't realize the mic would now work on the AE-5. It may not work consistently. I didn't really mess much with that. Maybe the new stuff added enough of a delay or something.

There are now controls for the ohm level on the headphones and the sound filter. And, the regular controls I added earlier for the Z. Equalizer stuff is able to be enabled with the variable definition for tuning, but the EQ adds some many extra volume level controls (11 IIRC) that it clutters up alsamixer. I think a separate application will have to be made for it.

**Comment by Conmanx360** [comment]
Have you tried disabling "Enable OutFX" and trying surround? Sound effects have some issues with surround.

Anyways, it probably isn't as simple as jack retasking. The ca0132 ignores most of the HDA standard once the DSP is loaded. I can look more into it in the morning. It's pretty late here.

**Comment by Jacek13** [comment]
Yes, I tried those already. thanks for so much work!

**Comment by Conmanx360** [comment]
What are you using to test surround? speaker-test -c6 is what I use, and I think that takes into account the ALSA chmaps. I don't think Pulseaudio does.

**Comment by dsteidley** [comment]
I have found speaker-test is the easiest to use.  Personally, I add the 
/-twav/ option.

David

**Comment by Conmanx360** [comment]
Something kind of interesting... so, when the mic isn't working, it seems like you can restore some input (although it's all messed up sounding, probably due to the sample rate being wrong) by setting paramID CONTROL_PARAM_VIP_SOURCE to 1. So, my guess is, it's some sort of audio stream mishap. There are streams internally that can be found in the 8051's exram (they start at 0x72f, and are ten bytes each). So, first step is understanding this stream interface. Then, moving to what value is being set incorrectly. I have dumps of the memory when it's working properly, so this should clear some stuff up.

Back to work I go.

**Comment by Conmanx360** [comment]
I've had to work quite a bit over the past couple of days, but I'm still making progress. Disassembly is a bit more difficult than just copying from dumps, but, it should give me the exact reason for the microphone failure, and then hopefully a way to get it working again. 

Creative got into contact with me about a week ago on the firmware stuff, and I asked them if they would consider releasing some sort of documentation. They still haven't gotten back to me on it, so I'm not sure if they're maybe discussing it internally or not. Either way, I'll continue the disassembly. If they decide to release documentation and help out, it would speed things up quite a bit.

**Comment by Conmanx360** [comment]
Ahh, found something interesting. The param names are in the 8051's program memory. So, that should help to clear up some of the undocumented params, although I don't think all of them.

**Comment by jon-jadot** [comment]
I can confirm Recon3di working.  At least headphone and line out on ga-x170-extreme-ecc.  I've been waiting a couple years for this.  Has worked since I installed kerenel 4.18.  Haven't tested other outputs as I have no use for them in my use case.

**Comment by dsteidley** [comment]
Ya, Conner has put a lot of work into this project and has even earned 
(in a good way) the attention of Creative Labs.  Of course, that isn't 
the only issue to make the "OS" decision.  One of the side benefits of 
working with him is that I now use VirtualBox to test out various 
distros and also test out installing software before I commit it to my 
actual environment.

**Comment by jon-jadot** [comment]
Qemu/KVM is a little more work, but worth the effort for testing etc.  As easy to setup when using a gui tool such as virt-manager.  VirtualBox is a little too generic for some testing.  I used qemu/kvm to passthrough my recon3di to windows while I waited for this lovely native linux driver.  At least I was able to use it in the meantime.  Whereas with virtualbox, regardless of what I tried, i couldn't get it to passthrough.

**Comment by Grief** [comment]
With the kernel 4.18.5 I have the only annoying issue with my Creative Sound Core3D built-in to ASRock Fatal1ty Z97 Professional. Microphone and speakers work just fine when connected to the rear green and pink outputs. But if I reboot to windows - the sound stops to work there and vise versa, when I reboot to linux from windows, sound is not working too. To fix that, I have to shutdown computer completely and plug the power off for some time. Looks like that the chip is saving its state even when computer is off but the power is connected. And looks like that windows and linux initializes the chip differently and cannot work with the chip if it was initialized by other os.

**Comment by jon-jadot** [comment]
Have you tried alsamixer to reinitialize in linux?   I've had a similar problem losing audio, but never had to shutdown for an extended period of time.  I used alsamixer to choose source out.  Then I would check pavucontrol to make sure app was outputting on correct channel as well.

**Comment by medzernik** [comment]
That bug happened to me back when the intel_hda_override=1 or something made my Recon3D work under Ubuntu 12.04.

**Comment by Conmanx360** [comment]
@grief -Yes, the onboard RAM of the DSP and it's 8051 processor do not clear with reboots, they need full shutdowns or a suspend. I have copied the shutdown procedures from within Windows though, so I don't think that's an issue for everyone. I did have reports of certain subsys ID Sound Blaster Z's having the same issue, too, even if I copied what Windows does on shutdown. My theory is, Windows does some sort of startup checks to fix this on subsequent reboots, but even if this is the case, it doesn't seem to work when going from Linux to Windows.

I've taken to just putting the computer to sleep to clear the card state if I need to. I understand it isn't ideal, but, it's what we've got to work with now. I think I'm beginning to understand the stream stuff that's handled by the 8051, and once I understand that, I should be able to diagnose any sort of no sound issue with dumps of it's memory. Even if I can't fix it on the fly, I should be able to get the DSP to redownload if the condition is met. This is done with the Sound Blaster Z, and it works for me, as I've never had sound not work since implementing it.

**Comment by Grief** [comment]
@jon-jadot sure. it doesn't help. Sometimes rmmod-insmod for snd_hda_codec_ca0132 helps but that's only in Linux. In Windows nothing helps, when the issue occurred first time and I didn't know the root cause of it, I did everything I could imagine, including driver reinstallation.

@Conmanx360 I see. Thanks for the clarification. Do you need some help? Maybe some donation?

**Comment by Conmanx360** [comment]
@Grief Well, donations are nice :) But I will continue to work on it even without donations. One thing that would be nice would be to get a board with a Recon3Di onboard. I have a 4790k and an i3 from that generation (I found them in the trash outside someones house, they do work though, funny enough), along with ram, so the cost would really just be the board. But all I've been able to find are ones that are still kind of expensive second hand. It's just easier to test on cards if I have them locally, speeds things up. That's really just a side thought. I might buy one anyways when I've got a bit more money.

Creative contacted me about two weeks ago on the whole firmware situation, but, they haven't responded back for some reason. I don't know anything about how companies function, but my hope is that they might be talking it over with some sort of legal or something, or maybe considering releasing the documentation. That's the hope. Either way, I intend on getting most of the issues fixed. 

Having the onboard 8051's program memory is extremely helpful, even if it is tedious. We've basically got the blueprint for how the thing works, it just needs to be disassembled and documented. I'm currently spending almost all the free time I've got and my work breaks disassembling and documenting it. So far, I've come to understand quite a few things better, and I think I'm getting close to understanding how to detect/possibly fix the microphone issue. The no sound issue MIGHT be a similar one from the dumps I have gotten, but I have no concrete way of knowing that yet. 

Really, once we come to understand the 8051, I think we will be definitively be able to solve problems without guessing work involved. It's just taking time.

**Comment by Grief** [comment]
@Conmanx360 please, provide us with PayPal or something else.

**Comment by Conmanx360** [comment]
I have it posted in a post from earlier that's a little higher in this chain, but I'll post it here again. I guess I need a more general place to put it.

[PayPal Link](https://www.paypal.me/conmanx360)

**Comment by medzernik** [comment]
Having Creative know about this entire thing makes me more nervous than happy. They are known for shutting down just about anything that they don't do themselves.

**Comment by Conmanx360** [comment]
@medzernik - There's really no risk of that, due to the way I reverse engineered the drivers. I haven't disassembled any binaries, I've only monitored communications between the device and the PC. This is how nouveau avoids any sort legal issues as well. The worst they could do is say "You can't distribute the firmware" which isn't the end of the world, because the Chromebook firmware does work. They seem like they're willing to help with the firmware. It's just taking them awhile.

It would be nice if they were more helpful, but I don't exactly need them to be either. With or without their help, it'll get fixed. It'd just make it easier if they would help out. Really, they'd be silly to come after me for this, because I've basically done a lot of work for them for free. They will probably make money from people actually buying their cards now that they have Linux support.

**Comment by medzernik** [comment]
@Conmanx360 There was a guy who basically fixed their drivers for Vista so people could use the card and he was prosecuted just because he did the job better than they did. They are c**ts. But hopefully the "directive" of Creative has changed - after all they are basically on the verge of bankrupcy.

**Comment by AvdWerfhorst** [comment]
@medzernik There are a few big differences. The guy did basically hack their property drivers, instead of writing something totally new. I think officially that was the problem. Also back then, they were trying to sell a newer sound card, so there was a direct financial gain by not fixing it. 
@Conmanx360 is writing a new driver by studying the behaviour of the communication between driver and hardware. So that is a totally different way of fixing it. One that is not using their code. And the cards are still top of their line. So it can only be positive for them, if Linux users are also starting to buy them.

**Comment by Conmanx360** [comment]
Still working on disassembly of the onboard 8051's program memory and mapping out ram locations and special function registers. Making progress, but it's pretty slow. If anyone else has knowledge or experience with 8051 disassembly and would like to help, let me know. I'm mainly interested in the stream stuff, but I am going ahead and documenting all the param functions to try and get some insight into things through them.

**Comment by Conmanx360** [comment]
Got a response from Creative finally, so, firmware stuff should be getting sorted soon. Seems like they might be interested in helping out with the mic issue, hopefully that works out. Will update here when I get more news on it.

**Comment by medzernik** [comment]
@Conmanx360 So what state is the AE-5 currently in, if you don't mind me asking?

**Comment by AvdWerfhorst** [comment]
@Conmanx360 How is your planning about submitting patches for getting this all upstream to the kernel?  Will you be waiting until the microphone is working, before submitting the patches. Or are there perhaps already a few new patches on their way to an upcoming kernel version?

**Comment by Conmanx360** [comment]
@medzernik - Full audio output support, inconsistent microphone input support. So, in my mind, pretty much fully functional minus the mic.

@AvdWerfhorst - I've kind of gotten side tracked on the whole microphone issue, as I kept getting complaints about it and it was really starting to bother me that I had everything BUT that working. But, I guess I should probably go about doing the driver cleanup, getting it into patches, and then go back to the microphone stuff. 

The main thing is, after getting a few complaints about the mic issue, I wasn't sure if people were even happy at all to get sound... The kind of feedback I was getting was "Oh. I have sound output. But my mic doesn't work." So, I got kind of sidetracked trying to fix that in hopes of avoiding those kind of comments in the future, but I guess most people will probably be happy to get sound out of their card at all like I was. I kind of feel bad releasing a set of patches when it doesn't consistently fully work, but I guess if the Sound Blaster Z is already like that, it won't cause any harm.

Hopefully that doesn't come across as whining or anything. I just want to release it and not have to get a ton of emails telling me that it doesn't work. I've thought about maybe setting up some sort of FAQ where I explain common issues and how to fix them, and detail things that are still not working. But, if you guys think it's a good idea to patch it up as is, and then submit the patches and do the microphone stuff later (which, is what I'm leaning towards, because mic stuff is really hard to put a time-table on) then that is what I will do. I think the AE-5 is definitely in a state that it could be submitted with minimal fixes. The ZxR will need some discussion with the alsa mailing list due to the weird nature of how I'm handling it.

Anyways, let me know what you guys think. Like I said, I'm already leaning towards splitting the AE-5 stuff into patches as is, and then there might be some delay between mic/ZxR patches.

**Comment by AvdWerfhorst** [comment]
@Conmanx360 I totally understand how the microphone gets you side tracked. And personally I am also someone who likes to see it work like it should, as it is a feature I use once in a while. Nevertheless, it's not the core functionality of the sound card. And I think a lot of people would be really happy to have their sound card output audio the way it is supposed to do. So in my opinion it is better to get the working parts first ready for patching into the kernel and after that focus on the microphone part. Of course there are some exceptions, but I think even the people who use the microphone, use it most of the time to output audio. To listen to music and other sounds.

**Comment by dsteidley** [comment]
@connor - I agree with the others that have chimed in and just want to 
re-add my "Thank you".  It has been great having the audio working on my 
system.  I do want the mic working... some day.  But I guess my 
suggestion would be to open up another repository on GitHub, just for 
the audio card project.  Then, we could start helping out with the side 
work that distracts you from the stuff you want to be doing.  We could 
be putting together the forum, the FAQs, and handling the bug reports.  
We could even help out with side projects like 8051 disassembly 
information gathering.  What do you think?

David

**Comment by Lolzen** [comment]
Barging in just to restate my gratitude.
I have the same opinion, that the milestone of worming sound shouldn't be held back for the mic.

Sure the mic not being fully there prevents some of my personal plans, however - that's my personal problem i have to and am dealing with.
I could imagine more folks actually using the sound output only, than people relying on mic, like i would. Streaming, playing etc. doesn't need input.

Having the patches upstream will give more people access to the state you've archived now, with a chunk of them probably not even aware of this github issue. Another chunk may not want to dive in to patch their kernel manually, so with that considered i have no doubt that milestones dhould be held back, because of pride/feelings etc. 
Sure, people will complain, we're on the internet after all.

Also, don't take feedback to issues necessarily in a bad way, just because - in my position - i can't help more than giving feedback, means i don't appreceate all the hard work, all these hours, even days i suppose, being worked into this.

**Comment by Conmanx360** [comment]
Thank you for the replies. Okay, it's good to hear people are fine with submitting in it's current form, then moving onto the mic and ZxR stuff after that. I don't mind bug reports, and it's good to get them. What kind of got me frantic about the mic stuff was a guy on Reddit who private messaged me about the AE-5 and wasn't too happy when I told him the mic didn't work. This is my first time really programming anything, so I'm still learning how to deal better with feedback, and should probably learn how to not let more negative comments bother me.

I'm kind of scattered all over the place now, from rushing to get sound out of the ZxR and AE-5. Then, the mic detour. So, I guess I should regroup and plan this out. First thing to do is separate the AE-5 stuff into patches. Then, I think I'll take @dsteidley 's suggestion and make a repository of my ca0132 tools, and the documentation on the 8051 that @abridgewater has written along with the stuff I've added. That way, more people can help out with the disassembly if they'd like to. Then, doing ZxR stuff, and after that I guess I will get back into the mic issue. If Creative is willing to help out, it may not end up being necessary to do all the 8051 disassembly stuff, but, it's probably good to get the information we've gathered so far out there even if it isn't needed for anything immediately.

Alrighty, with that out of the way, getting to work on it.

**Comment by medzernik** [comment]
> @medzernik - Full audio output support, inconsistent microphone input support. So, in my mind, pretty much fully functional minus the mic.
> 
> @AvdWerfhorst - I've kind of gotten side tracked on the whole microphone issue, as I kept getting complaints about it and it was really starting to bother me that I had everything BUT that working. But, I guess I should probably go about doing the driver cleanup, getting it into patches, and then go back to the microphone stuff.
> 
> The main thing is, after getting a few complaints about the mic issue, I wasn't sure if people were even happy at all to get sound... The kind of feedback I was getting was "Oh. I have sound output. But my mic doesn't work." So, I got kind of sidetracked trying to fix that in hopes of avoiding those kind of comments in the future, but I guess most people will probably be happy to get sound out of their card at all like I was. I kind of feel bad releasing a set of patches when it doesn't consistently fully work, but I guess if the Sound Blaster Z is already like that, it won't cause any harm.
> 
> Hopefully that doesn't come across as whining or anything. I just want to release it and not have to get a ton of emails telling me that it doesn't work. I've thought about maybe setting up some sort of FAQ where I explain common issues and how to fix them, and detail things that are still not working. But, if you guys think it's a good idea to patch it up as is, and then submit the patches and do the microphone stuff later (which, is what I'm leaning towards, because mic stuff is really hard to put a time-table on) then that is what I will do. I think the AE-5 is definitely in a state that it could be submitted with minimal fixes. The ZxR will need some discussion with the alsa mailing list due to the weird nature of how I'm handling it.
> 
> Anyways, let me know what you guys think. Like I said, I'm already leaning towards splitting the AE-5 stuff into patches as is, and then there might be some delay between mic/ZxR patches.

The full audio support is phenomenal. The mic not working is something that personally annoys me - but I think that is more because of the nature of soundcards - many people just need all the features of the soundcard, not just partial ones, so that's why everyone is pointing out something.

The fact that the card outputs sound now (and I presume it's configurable) is just amazing. I am so looking forward once the mic works to get a proper EQ/nice sound working in Linux. I am a bit tight on money right now, but when it's all done, expect a donation :D

**Comment by Conmanx360** [comment]
Okay, got today and tomorrow off, gonna try to get some more work done on it. I've been working a little slow due to working quite a bit. I'm off Monday-Thursday next week, so hopefully I get it split into patches by then.

Also, Creative gave me the stuff needed to submit the firmware, so I've got to get that done, too.

**Comment by ThomasAH** [comment]
linux-image-4.18.0-0.bpo.1-amd64 (4.18.6-1~bpo9+1) is now available in Debian stretch-backports, so installing it in stretch became really easy.

Thank you for your work!

**Comment by Conmanx360** [comment]
Okay, about halfway through getting the AE-5 stuff into patches. Hoping to get it done by tonight. If not, it should be done by tomorrow.

This is significantly more code than the Recon3D stuff, so it may require some back and forth. Then, I'll be onto getting the firmware sent off.

**Comment by Conmanx360** [comment]
Okay, got the AE-5 patches sent. Now, time to see about the firmware stuff.

I changed the file ctefx-sbz.bin to ctefx-desktop.bin because all the desktop cards seem to use it, so it was no longer Sound Blaster Z specific. So, if you use the patch from the email, be sure to change ctefx-sbz.bin to the new name.

**Comment by Conmanx360** [comment]
Okay, looks like the patches were accepted. Now, time to handle the firmware stuff and figure out how to get the ZxR into a cleaner state.

**Comment by Conmanx360** [comment]
Sorry for the lack of updates, been busy over the past few days. Anyways, I've now got the ZxR's digital input working, which is nice. I'm close to having it in a state where I believe it can be submitted to the kernel. 

I'm treating the DBPRO daughterboard as basically a separate, generic card that only has digital in/out for now. I will see if I can't get it's line-in working without it being too much of a hassle, it kind of gets confusing when they're treated as one in alsa-mixer...

Anyways, like I said, making progress. :)

**Comment by dsteidley** [comment]
Sweet!  Thanks again for all of the hard work.  As always, let me know 
if you need a hand with any documentation.  I check in to the Github 
site to see if there is anything I can do.


David

**Comment by Conmanx360** [comment]
@dsteidley - No problem. I plan on setting up that repository once I've got all this stuff sorted, although I guess I could set it up earlier, depending on how long this takes.

I've got the analog aux-in on the daughterboard of the ZxR working, so that means all of it's features (Digital In/Out, Aux-in) are working. I'm in the process of splitting it into patches right now, but since I work this weekend, it may not get done until Sunday or Monday. I've also added patches to fix the surround sound with output effects enabled having improper channel mixing.

On a recent bootup with the ZxR, the microphone and the input effects were all working. This leads me to believe that my commands aren't wrong, but the timing is. I may have already mentioned this earlier. Anyways, Once these ZxR patches are submitted, the plan is to add the ability to see the timing information from qemu to QemuHDADump, so that I can see if there are any obvious delays between commands in the microphone setup process. Fingers crossed that this is the issue, and that further 8051 disassembly isn't required, because that's pretty time consuming.

Anyways, like always, will update with progress here.

**Comment by Conmanx360** [comment]
Okay, just finished the ZxR patches. Gotta go to work now, but I'll do another look over them and then hopefully get them sent out tonight. That should wrap up getting audio output on the PCI-E Creative cards, next goal is the microphone.

**Comment by Conmanx360** [comment]
Got the ZxR patches sent out on Saturday night, looks like they should be good. Turning focus back on microphone now, already got timing info from QemuHDADump, so hopefully that should give me enough info to figure it out.

**Comment by Lolzen** [comment]
This might be somewhat related:
As i'm dual booting with Win7 currently, i noticed when i didn't fully power down the PC after being in Linux the mic loudness is affected in Win. 
More precise: with a patched kernel the mic is significant more down in input volume, despite no settings changed until shutdown, boot. With an unpatched kernel (i.e. no sound) the mic is not picking up anything without proper shutdown. Reeboots don't "reset" the card.

Might have nothing to do with it at all, just though i'll throw it in here so it could be somewhat helpful(?)

**Comment by Conmanx360** [comment]
Yeah, the original ca0132 driver (unpatched one) sets some stuff up that is unfixable without a full shutdown and clearing of the onboard memory. If I had documentation, I could probably figure out what is being set and un-set it to fix it... might even be related to the fact that the original ca0132 driver tries setting up a digital mic, which the desktop cards don't have.

I'm kind of at the end of my wits with the microphone. Not many more ideas for fixing it. I still have a few, but I feel like they're not very likely to work. Once again, it's another case of lacking the proper documentation to actually debug what's going wrong. I believe that all of the info could be found through disassembly of the 8051's program, but like I've said before, that could take quite a bit of time. 

I know I'm sending the exact same commands that Windows is, which to me seems to say that something else is wrong, either with the way Linux sets up streams, or maybe setting up controls causes it to run commands more than once, I really don't know. I know that my earlier drivers had the Sound Blaster Z microphone working all the time, and input effects worked fine as well. Maybe I should roll back to those and get some dumps of what verbs are sent from a virtual machine, because doing diffs between that driver and the ones where it began to break gave no indication on what went wrong. 

I'm sure I'll figure it out. It's just going to take some time.

**Comment by Lolzen** [comment]
Thanks for the insight!
It's not a very urgent problem at all imo, like i said i just wanted to be sure you're aware of it, which you just confirmed :)

**Comment by Conmanx360** [comment]
@Lolzen - No updates on that. Not sure what to expect. I was told that they were looking into ways that they could help, but nothing more than that. So, I think it's best to expect that they most likely won't be helping out much on the technical side.

I never really got my hopes up on it, although it did sound good. I'll continue to bug them and see what I get out of it.

**Comment by Conmanx360** [comment]
Ahhh, this microphone issue is exhausting. Not even sure I'm making progress. Not sure if I mentioned it before, but it used to work perfectly fine in an older version of the driver I had written which only had stuff for the Sound Blaster Z. I've moved back to that version, and I'll be trying to debug the execution of it vs. the new, non-working one. So, hopefully now I'm not just trying things without any idea of what I'm doing.

Creative Labs now has a [subreddit](https://www.reddit.com/r/SoundBlasterOfficial/). I asked the moderator if I could make a post there on this, and he said that it'd be fine. So, I may make a post over there. It'd probably include a FAQ of some sort, and the stuff to figure out DKMS, instead of having it on some Github project that may end up being hard to find... I'll update here if/when I end up making a post there.

**Comment by Conmanx360** [comment]
Soooooooooooooo, seems like I solved the mic issue. Funny enough, a bug in the earlier versions of the code was actually what made it work. I don't know why it did, but it did, and now the microphone seems to be working fine as well as the input effects. 

I'll try to mess with it a bit and see if I can't make it look better. If you are compiling your own patch and want to fix it, for Sound Blaster Z/ZxR, in sbz_setup_defaults, before the dsp download check, add:
```
        if (spec->quirk == QUIRK_SBZ || spec->quirk == QUIRK_ZXR)
		ca0132_alt_dsp_scp_startup(codec);
```

Similarly, for the AE-5, in ae5_setup_defaults, add:
```
        if (spec->quirk == QUIRK_AE5)
		ca0132_alt_dsp_scp_startup(codec);
```

Like I said, this is kind of silly. The checks for the quirks might not be necessary, but I haven't tested it out without them yet, and it works like that as for now. 

If anyone else tests it, let me know if it works for you. After so many attempts at fixing it, I have my doubts, but so far it's been working. Now that that issue is (seemingly) out of the way, I feel a lot better.

**Comment by Conmanx360** [comment]
Okay, looks like an 

`msleep(100);`

at the start of the setup_defaults function is sufficient enough to make the microphone work, and that the commands only served as a delay. Should be a pretty simple patch. Will test it out for a little bit before sending it out.

**Comment by Conmanx360** [comment]
Sent the microphone patches out, so, that should be it. Sending the firmware is the next on my list of things to do, then I'm pretty sure that wraps up getting the cards working for the most part.

Anyone know of anything else that still needs done? I can't really think of anything, I guess the AE-5 LED stuff, but that's not really necessary to get the card working or anything.

**Comment by nikosmar** [comment]
Since this is my first post here, I have to thank you for your work.

> Anyone know of anything else that still needs done? I can't really think of anything, I guess the AE-5 LED stuff, but that's not really necessary to get the card working or anything.
I have SBZ on (K)Ubuntu 18.04.1 kernel 4.18.10 and sometimes (I can't remember exactly when) I have no sound after rebooting. This is fixed by toggling line out/headphones on alsamixer, not a serious bug of course.
I think I have read it already, but I am not sure if it's fixed.

**Comment by Conmanx360** [comment]
@nikosmar - Hmm, that's a weird bug I've never had / heard anyone mention. Maybe it has something to do with your alsamixer controls not being saved or something? What all do you have plugged in? Headphones and line-out, or just headphones?

**Comment by Conmanx360** [comment]
Okay, I created a thread over on the [SoundBlasterOfficial subreddit](https://www.reddit.com/r/SoundBlasterOfficial/comments/9mm5ad/sound_blaster_r3dr3dizzxrae5_linux_driver/), so migrating over there would probably be a good idea. I'll still try to respond to issues here, but it'd probably be easier to do any discussion over there.

**Comment by nikosmar** [comment]
@Conmanx360 I have plugged in my headphones, the card's mic and a set of logitech z506. I will write down when does this happen, and try to reproduce it, then I'll post my results on the reddit thread.

**Comment by lss4** [comment]
I've been thinking about posting this minor issue for a while... just not sure if the issue is with the driver or with PulseAudio. I'm using Sound Blaster ZxR with latest Manjaro Cinnamon on 4.20 kernel.

It seems for some reasons, with the desktop environment's Sound Settings (it uses PulseAudio), both Headphones and Line Out outputs to speakers. Actually, the underlying alsamixer settings set output to Line Out on both output options. I can manually set the output to Headphones using alsamixer, but it won't survive across reboots which means I need to manually change this every startup. While this can be a little inconvenient, everything essential works, including mic and speaker/headphone output.

Manjaro Cinnamon is the only OS installed on this system, so issues related with dual-booting is not involved.

**Comment by Jacek13** [comment]
@lss4 Destktop sound settings don't work for me either, no matter the OS.
I just use alsamixer instead and run it with sudo (`sudo alsamixer`) to make the settings stick after reboot. I also read that when even that doesn't work, you should use `sudo alsactl store`.
I use Sound Blaster Recon 3Di (Sound Core 3D - integrated into the mobo) for future refference.
Good luck!

**Comment by lss4** [comment]
The issue with desktop sound settings (PulseAudio) is still here.

While we can work around by using alsamixer (bypassing PulseAudio), it's still kinda inconvenient when we can't just use our desktop environment's sound settings...

**Comment by TTimo** [comment]
Since there is no clear problem with SteamOS here, but rather a problem with Linux support in general for some of Creative Labs's products, I'm going to be closing this report. Please open a different report if you have compatibility problem specifically with SteamOS. See the links in the history to SoundBlasterOfficial's subreddit, etc.

**Comment by iblender** [comment]
@Conmanx360 I put the files into the folders as said. I then clicked on my soundblaster zxr and clikced '5.1' and then the device disappeared. Still cant hear anything. I know this thread is super old but if anyone is still around could you help out please?
